File: solutions/clojure/matching-brackets/1/src/matching_brackets.clj
(ns matching-brackets)

(def ^:private openers #{\( \[ \{})
(def ^:private closers {\) \(, \] \[, \} \{})

(defn valid? [s]
  (= []
     (reduce (fn [stack ch]
               (cond
                 (openers ch) (conj stack ch)
                 (closers ch) (if (= (closers ch) (peek stack))
                                (pop stack)
                                (reduced nil))
                 :else stack)) [] s)))

File: solutions/ruby/matching-brackets/1/matching_brackets.rb
module Brackets
  def self.paired?(str)
    stack = []
    str.each_char do |ch|
      stack << ')' if ch == '('
      stack << ']' if ch == '['
      stack << '}' if ch == '{'
      return false if ')]}'.include?(ch) && stack.pop != ch
    end
    stack.empty?
  end
end
